### PR TITLE
Chronos : support confidence interval for forecaster

### PR DESCRIPTION
**Pull request by rnwang04**
## Description

Confidence interval to be supported in forecaster (TCN, S2S, NBeats, LSTM, Autoformer).

### 1. Why the change?

To provide confidence interval in forecaster for user.

### 3. Summary of the change 

Confidence interval to be supported in forecaster (TCN, S2S, NBeats, LSTM, Autoformer).

### 4. How to test?
- [x] Unit test
- [x] Jenkins
http://10.112.231.51:18889/view/BigDL-PR-Validation/job/BigDL-Chronos-PR-Validation/862/

### CODE REVIEW

File: python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py [hunk 1]
```diff
@@ -779,6 +779,146 @@ def evaluate_with_onnx(self, data,
...
+                self.internal.eval()
+                yhat = _pytorch_fashion_inference(model=model,
+                                                  input_data=data,
+                                                  batch_size=batch_size)
+                if not is_local_data:
+                    yhat = np_to_xshard(yhat, prefix="prediction")
+                return yhat
+
+        # step3: calculate y_hat
+        y_hat = calculate(data, self.internal)
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py:899]: should we use the mean y_hat as the final result?

File: python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py [hunk 2]
```diff
@@ -779,6 +779,146 @@ def evaluate_with_onnx(self, data,
...
+        # which should be done during fit
+        if not hasattr(self, "data_noise"):
+            invalidInputError(validation_data is not None,
+                              "When call predict_interval for the first time, you must pass in "
+                              "validation data to calculate data noise.")
+            invalidOperationError("mse" in self.metrics,
+                                  "You must add 'mse' in metrics when initialize Forecaster.")
+            output = self.evaluate(data=validation_data, multioutput='uniform_average')
+            index = self.metrics.index("mse")
+            self.data_noise = output[index]
```
[the remark this answers is not in the document]
Comment (TheaperDeng) [reply python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py:851]: and btw, the data_noise is calculated as a "uniform average", which means it is a scaler

File: python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py [hunk 3]
```diff
@@ -779,6 +779,146 @@ def evaluate_with_onnx(self, data,
...
+            y_hat_list.append(calculate(data, self.internal))
+        y_hat_mean = np.mean(np.stack(y_hat_list, axis=0), axis=0)
+        invalidInputError(y_hat_mean.shape == y_hat.shape,
+                          "dismatch shape between y_hat_mean and y_hat")
+
+        model_bias = np.zeros_like(y_hat_mean)
+        for i in range(repetition_times):
+            model_bias += (y_hat_list[i] - y_hat_mean)**2
+        model_bias /= repetition_times
+        std_deviation = np.sqrt(self.data_noise + model_bias)
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py:918]: and here we have model_bias to be a 3dim vector.

File: python/chronos/test/bigdl/chronos/forecaster/test_tcn_forecaster.py [hunk 4]
```diff
@@ -678,3 +678,55 @@ def test_tcn_forecaster_tune_save_load(self):
...
+                                   output_feature_num=1,
+                                   kernel_size=4,
+                                   num_channels=[16, 16, 16],
+                                   loss="mse",
+                                   metrics=["mae"],
+                                   lr=0.01)
+        forecaster.fit(train_data, epochs=2)
+        with pytest.raises(RuntimeError):
+            y_pred, std = forecaster.predict_interval(data=test_data[0],
+                                                    repetition_times=5)
```
Comment (TheaperDeng) [inline python/chronos/test/bigdl/chronos/forecaster/test_tcn_forecaster.py:732]: shall we add some test for distributed forecaster?

File: python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py [hunk 5]
```diff
@@ -779,6 +779,146 @@ def evaluate_with_onnx(self, data,
         aggregate = 'mean' if multioutput == 'uniform_average' else None
         return Evaluator.evaluate(self.metrics, target, yhat, aggregate=aggregate)
 
+    def predict_interval(self, data, validation_data=None, batch_size=None, repetition_times=5):
+        """
+        Calculate confidence interval of data based on Monte Carlo dropout(MC dropout).
```
[the remark this answers is not in the document]
Comment (rnwang04) [reply python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py:784]: done

File: python/chronos/test/bigdl/chronos/forecaster/test_tcn_forecaster.py (hunk 4 above)
Comment (rnwang04) [reply python/chronos/test/bigdl/chronos/forecaster/test_tcn_forecaster.py:732]: yes, will add more ut for different data types later

File: python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (rnwang04) [reply python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py]: Good catch, will use predict + Evaluator.
will change data_noise to 3dim vertor as mean_y_hat

File: python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py (hunk 1 above)
Comment (rnwang04) [reply python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py:903]: yeah, will use mean y_hat

File: python/chronos/test/bigdl/chronos/forecaster/test_tcn_forecaster.py (hunk 4 above)
Comment (rnwang04) [reply python/chronos/test/bigdl/chronos/forecaster/test_tcn_forecaster.py]: After discussion with @TheaperDeng , will not support distributed forecaster in predict_interval.

File: python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py [hunk 6]
```diff
@@ -779,6 +779,137 @@ def evaluate_with_onnx(self, data,
...
+
+        if not self.fitted:
+            invalidInputError(False,
+                              "You must call fit or restore first before calling predict_interval!")
+
+        def calculate(data, model):
+            yhat = _pytorch_fashion_inference(model=model,
+                                              input_data=data,
+                                              batch_size=batch_size)
+            return yhat
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py:850]: have no objection about this, but why we need this temp function here, any specific reason?

File: python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py [hunk 7]
```diff
@@ -779,6 +779,137 @@ def evaluate_with_onnx(self, data,
...
+
+        # turn on dropout
+        self.internal.apply(apply_dropout)
+
+        y_hat_list = []
+        for i in range(repetition_times):
+            y_hat_list.append(calculate(data, self.internal))
+        y_hat_mean = np.mean(np.stack(y_hat_list, axis=0), axis=0)
+        invalidInputError(y_hat_mean.shape == y_hat_list[0].shape,
+                          "dismatch shape between y_hat_mean and y_hat")
```
Comment (TheaperDeng) [inline python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py:903]: why we need this invalidinputerror check?

Comment (rnwang04) [reply python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py:903]: will delete this check.

File: python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py (hunk 6 above)
Comment (rnwang04) [reply python/chronos/src/bigdl/chronos/forecaster/base_forecaster.py:850]: just avoid to write this content two times. can remove this temp function.

**Pull request merged**